app/cdap/components/DataPrep/Directives/Calculate/index.js

```diff
@@ -466,27 +491,37 @@ export default class Calculate extends Component {
 
   parseColumns = () => {
     let columns = typeof this.props.column === 'string' ? [this.props.column] : this.props.column;
+    let columnTypes = new Set(columns.map(col => DataPrepStore.getState().dataprep.typesCheck[col]));
 
     let type1 = DataPrepStore.getState().dataprep.typesCheck[columns[0]];
     if (columns.length === 1) {
       return {
         isDisabled: false,
         crossColumn: false,
+        multiColumn: false,
         columns: columns[0],
-        columnType: type1
+        columnTypes: [...columnTypes],
+        areColsMixedTypes: false
       };
     }
 
-    if (columns.length === 2) {
-      const type2 = DataPrepStore.getState().dataprep.typesCheck[columns[1]];
-      let index = columns.indexOf(this.props.ddSelected);
-      if (type1 === type2 && index !== -1) {
-        columns = index === 0 ? columns : [this.props.ddSelected, columns[0]];
+    if (columns.length >= 2) {
+      if (columns.includes(this.props.ddSelected)) {
+        // Add currently-selected column to front of column list
+        columns = columns.filter(col => col !== this.props.ddSelected);
+        columns.unshift(this.props.ddSelected);
+
         return {
-          isDisabled: false,
-          crossColumn: true,
+          /*
+          Disable Calculate if both numerical and String-type columns are selected.
+          This could be changed in the future if some options can operate on both at once.
+          */
```

Comment:
Can this be detected by the type restrictions of the directives? I.e. if no directives are available, disable the menu?

Reply:
I thought this wasn't possible at first, but it ended up just needing a refactoring of `parseColumns`. It looks a lot different now, let me know if you prefer it this way or how it was before.